Add search and zip features

kabirtomer opened the pull request:
This is to add the implementation of 
- Change in the database. All approved files are now stored in media/files, with a .meta file in their place containing within the title the filename and a link to the actual file, and storing the file's meta tags for search purposes.
- Interactive Search to the frontpage. 
- Filename Search to the approvals page.
- Finalize Approvals link that zips and stores courses.

ShouryaAggarwal committed:
Added zip functionality to each approval

ShouryaAggarwal committed:
ignored .zip and .meta in database.json

kabirtomer committed:
Add JSON reading functionality at frontend

kabirtomer committed:
Add keyword search for dictionaries

kabirtomer committed:
Add search API

kabirtomer committed:
Display search results

kabirtomer committed:
Fix incorrect rendering of search results

kabirtomer committed:
Merge search and browse pages

ShouryaAggarwal committed:
Improved zip and metadata features

ShouryaAggarwal committed:
Changed readme

ShouryaAggarwal committed:
Added the functionality of finalizing approvals

ShouryaAggarwal committed:
Made zipping feature more efficient

ShouryaAggarwal committed:
Added a function to build up metadata for pasted files

kabirtomer committed:
Add text-based search ranking

kabirtomer committed:
Merge with Zip feature branch

kabirtomer committed:
Fix search on metafile names

kabirtomer committed:
Add metadata search

ShouryaAggarwal committed:
Merged the changes facilitating search

kabirtomer committed:
Import HttpResponse and remove recreate_path from remove_unapproved_document

ShouryaAggarwal committed:
Fixed finalize approvals function

ShouryaAggarwal committed:
Fixed finalize approvals

kabirtomer committed:
Fix whitespace in meta tags and remove debug code

ShouryaAggarwal committed:
Added functionality to record zip dl stats and to remove them

kabirtomer committed:
Add rough file search to approvals page

kabirtomer committed:
Comment search code

ShouryaAggarwal committed:
Commented zip and metadata generation features

ShouryaAggarwal committed:
removed .idea

kabirtomer committed:
Merge zip comments and tracking model

ankurshaswat commented:
Create a script to make the required folders.
Commenting is not standaradized. 3 types of commenting used // , ''' (Single line) , ''' (Multiline)

aagrxyz commented:
The code is broken. 
Cloning `kabir:master` and running `python manage.py runserver` gives an import error.
` File "/tmp/BookShelf/bookShelf/books/urls.py", line 3, in <module>
    from . import views
  File "/tmp/BookShelf/bookShelf/books/views.py", line 22, in <module>
    from books import search as search
  File "/tmp/BookShelf/bookShelf/books/search.py", line 1, in <module>
    from . import views
ImportError: cannot import name 'views'`

aagrxyz reviewed (changes requested):
Function APIsearch is repeated in views.py 
Check lines 265 to 300

aman71197 commented:
There are multiple places where file paths have been joined manually with just string concatenation.
Change all of them to os.path.join
This is necessary to make it platform independent and for an error-free code

aagrxyz reviewed (changes requested):
There are multiple places where file paths have been joined manually with just string concatenation.
Change all of them to os.path.join
This is necessary to make it platform independent and for an error-free code

kabirtomer committed:
Standardize commenting

kabirtomer committed:
Fix circular import

kabirtomer committed:
Make path join OS independent

aagrxyz reviewed (changes requested):
course_downloads.txt is not created automatically the first time. Make sure you create the file in the necessary format if it is not present.

[Err] No such file or directory: 'course_downloads.txt'
Request Method: | GET
http://localhost:8000/books/download_course/?course=COL106

kabirtomer committed:
Add course_downloads.txt and script to correctly place/replace it

aagrxyz commented:
Presently the media/files directory just contains the dump of files. So now let's say multiple courses have a file known as lec1.pdf then the files will be overwritten.  I added two different files, but both named as lec01.pdf to COL202/Professors/amitk/ and COL352/Professors/naveen/
Finally, after Finalize Approvals, both the lec01.pdf in COL202 and COL352 pointed to the file lec01.pdf in the COL352 directory.
This seriously limits the names, that we can give to files. This should not be the case.

aagrxyz reviewed (changes requested):
Overwriting of files with same name,

kabirtomer committed:
Fix fifilename clashes and add rewrite warning

kabirtomer committed:
Fix upload filename clash

aagrxyz commented on bookShelf/books/views.py:
```diff
@@ -17,147 +19,435 @@
+
+	with open(STATS_FILE, "w") as stats:
+		stats.writelines(lines)
+	return redirect('/../media/database/' + parent_dir + '/' + course + '.zip')
```
Use the declared global variables to get the file. Don't Hardcode

aagrxyz commented:
When the admin, clicks to view a file the browser treats it as a Binary file, as it doesn't have an extension.
Like MT$MTL100$Professors$atripath$Notes$[Spring18]-MTL100-Notes-random.pdf==MTL100$atripath
If you click, to see its content, it is treated as a binary file.

aagrxyz commented on bookShelf/books/views.py:
```diff
@@ -17,147 +19,435 @@
+				while os.path.exists(temp_path) and os.path.isfile(temp_path):
+					temp_path=file_path+'('+str(i)+')'
+					i+=1
+				file_path=temp_path  
```
File path doesn't have an extension, causing it to act as a binary file.

aagrxyz commented on bookShelf/books/views.py:
```diff
@@ -17,147 +19,435 @@
+								path = os.path.join(dirname, filename)
+								to_write = os.path.join(os.path.split(os.path.relpath(path, course_path))[0],
+														os.path.split(file_loc)[1])
+								zf.write(file_loc, arcname=to_write)
```
This causes file not found error, when clicking on finalize approvals, in a course that just got its first file. This is because the meta file exists but it's corresponding file does not exist in the FILESV_DIR

aagrxyz reviewed (changes requested):
There is still conflict between two files having the same name when they are pasted manually. We need to work a better way to store the meta data. Having just a single dump of files has a lot of problems.

ShouryaAggarwal committed:
Fixed minor Issues

ShouryaAggarwal committed:
Fixed directory structure, improved zip feature and bulk approvals, earlier creation of meta files, force integrity feature, added journalling feature (partial)

ShouryaAggarwal committed:
Reverted naming of uploaded files back to minimal

ShouryaAggarwal committed:
Made a startup function to execute task_file and check jsons

ShouryaAggarwal committed:
Updated readme

aagrxyz commented on bookShelf/books/views.py:
```diff
@@ -17,147 +19,422 @@
 def remove_unapproved_document(request):
+	name = request.GET.get('name', 'none')
+	metafile = name + '.meta'
 	try:
```
Do this in 2 separate try catch blocks. It is possible that first one causes error and so .meta file will bot be deleted

aagrxyz replied:
@kabirtomer Look into this

aagrxyz commented:
Change The structure of task_file.json
Presently you need to iterate over all courses once in removal and once in additions
Instead store the structure as


`[
    { course: "COLxxx" , change:" something" , type: "addition" , timestamp: "xxxxx"} ,
   { course: "COLxxx" , change:" something" , type: "removal" , timestamp: "xxxxx"}
]`
Thus store each addition and removal as a list of actions that need to be performed on it with the timestamp. This is necessary as there is an order in the actions performed by the user and we don't want to have an arbitrary order on the actions.

ShouryaAggarwal committed:
Changed structure of task_file.json

ShouryaAggarwal committed:
Modified remove_unapp_docs function to check for separate files first

kabirtomer committed:
Add function to update database.json using task_file.json

kabirtomer committed:
Fix UI for tree based storage

kabirtomer committed:
Add metatag e edit feature to rename and fix search

ShouryaAggarwal committed:
Merged remote changes

aagrxyz commented:
Rename button at Admin panel is not working
Got the following error: 
NameError at /books/rename/
name 'name' is not defined

aagrxyz commented on bookShelf/books/views.py:
```diff
@@ -17,147 +19,494 @@
+		task[COURSE] = separated_path[1]
+		task[TYPE] = type_of_change
+		task[PATH] = path
+		task[TIMESTAMP] = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
```
Add timezone to this. Make sure this timestamp is  in GMT+5:30

ShouryaAggarwal committed:
Modified delete_zips.py

aagrxyz commented:
1. Upload Form Does not have an option to fill in meta data tags
2. Whenever the user downloads the file, the file is saved with name as "download". Need to send some header that states the name of file to be stored when downloaded.
3. Remove empty folders from bulk directory after you have moved files to the database directory

aagrxyz committed:
Update shelf.html

kabirtomer committed:
Fix renaming

kabirtomer committed:
Correct downloaded file name

kabirtomer committed:
Modify delete_zips.py

kabirtomer committed:
Correct Time Zone

kabirtomer committed:
Improve tagging UI

ShouryaAggarwal committed:
Added changes to restore bulk_dir

aagrxyz committed:
added invalid course code error checking in download course function and removed heartbeat api

Commit:
added invalid course code error checking in download course function and removed heartbeat api

Commit:
Merge branch 'master' of https://github.com/kabirtomer/BookShelf

aagrxyz committed:
Added some error checking in functions.
Removed search bar in approve html page
Decoupled some repetitive code into functions

aagrxyz commented:
I have made some changes in the code. The code related to search feature is still left to be reviewed.
@mayanksingh2298 Go through all the code once and update your progress here.
Also @kabirtomer , a lot of work needs to be done at the UI side. Make sure that is done.

ShouryaAggarwal committed:
Cleaned some code

kabirtomer committed:
Add user provided metatag functionality

kabirtomer committed:
Fix reload tags bug

ShouryaAggarwal committed:
Minor changes in UI

kabirtomer committed:
Readd file search to approvals page

kabirtomer committed:
Fix search and button classes and UI

ShouryaAggarwal committed:
Added warning for pending finalize approvals

ShouryaAggarwal committed:
Merge branch 'master' of https://github.com/kabirtomer/BookShelf

Merged remote changes

kabirtomer committed:
Fix front page box sizes

kabirtomer committed:
Correct search bar styling

kabirtomer committed:
Make search bar UI consistent

kabirtomer committed:
Correct searcrcha placeholders

kabirtomer committed:
Merge remote-tracking branch 'origin/master' into UI

ShouryaAggarwal committed:
Fixed approve page minor errors

ShouryaAggarwal committed:
Added cronjob to delete zips

kabirtomer committed:
Clean up upload UI

ShouryaAggarwal committed:
Added url encoding on approve page links

kabirtomer committed:
Merge branch 'master' of github.com:kabirtomer/BookShelf

kabirtomer committed:
Fix upload UI

mayanksingh2298 commented:
![image](https://user-images.githubusercontent.com/25145295/41432481-e704b896-7033-11e8-9b4c-927da29c8dfd.png)
"../" is not required in the highlighted of Readme.md
![image](https://user-images.githubusercontent.com/25145295/41432609-4225e52e-7034-11e8-93ee-44267a7035bf.png)
Also you need a "cd .." before heading in the bookshelf folder

mayanksingh2298 commented:
![image](https://user-images.githubusercontent.com/25145295/41432917-09e869c4-7035-11e8-961b-b4d0ebe66055.png)
remove  "(Create a super user by following the instructions)"
The second last line gives an error:
![image](https://user-images.githubusercontent.com/25145295/41432953-2383e480-7035-11e8-825c-cf865d3081ef.png)

kabirtomer committed:
Fix tag UI

kabirtomer committed:
Fix approvals UI

kabirtomer committed:
Make searchbars responsive

kabirtomer committed:
Fix Upload form validation

aagrxyz commented:
![image](https://user-images.githubusercontent.com/18374988/41456513-a8882f7a-70bb-11e8-9712-081f56045b54.png)
1. Why is there so much white space on the left,right and bottom margin. Make use of every space on the screen and increase the size of the div.
2. To search a file, I need to manually click on the search button. Add an event listener for enter key in that form.
3. Clicking on reset buttons does not clear the search box.

![image](https://user-images.githubusercontent.com/18374988/41456649-18cfd526-70bc-11e8-88c9-f5a329194a93.png)

4. Make the "NEXT" button blue in color (by default, not only on hovering the mouse) to match our color scheme and put it in the center of the div, and reduce it's width. Do the same with "SUBMIT" button later on.

And good work on that tag form, it looks amazing.

kabirtomer committed:
Make search bars responsive to enter key

ShouryaAggarwal committed:
Added backend form validation

kabirtomer committed:
Add smooth scaling UI for larger screens

ShouryaAggarwal committed:
Modified readme

ShouryaAggarwal committed:
Modified readme

ShouryaAggarwal committed:
Made Changes to cronjob

ShouryaAggarwal committed:
Fixed backend form validation

ShouryaAggarwal committed:
Modified backend form validation

ShouryaAggarwal committed:
Modified backend form validation

ShouryaAggarwal committed:
Added regex test to validator

ShouryaAggarwal committed:
Minor changes

kabirtomer committed:
Change form buttons to match with theme

aagrxyz committed:
Removed circular imports and changed encoding of requirements.txt to UTF-8

aagrxyz committed:
Update README.md

kabirtomer committed:
Scale content height for large screens and style download button

kabirtomer committed:
Merge branch 'master' of https://github.com/kabirtomer/bookShelf

ShouryaAggarwal committed:
Minor bug fixes

kabirtomer committed:
Minimize wasted space for all screen sizes

kabirtomer committed:
Merge branch 'master' of https://github.com/kabirtomer/bookShelf

ShouryaAggarwal committed:
Minor changes to download_zip module and removed some unused static files

ShouryaAggarwal committed:
removed test files

ShouryaAggarwal committed:
added missing file
